class Solution {
    public int numDistinct(String s, String t) {
        int[][] dp = new int[s.length()][t.length()];
        for(int i = 0 ; i<s.length() ; i++){
            Arrays.fill(dp[i],-1);
        }
        return solve(s,t,0,0,dp);
    }

    public int solve(String s , String t , int i , int j , int[][] dp){
        if(j==t.length()){
            return 1;
        }
        if(i>s.length()-1){
            return 0;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        int include = 0;
        int exclude = 0;
        if(s.charAt(i) == t.charAt(j)){
            include = solve(s,t,i+1,j+1,dp);
            exclude = solve(s,t,i+1,j,dp);
            return dp[i][j] = include + exclude;
        }
        return dp[i][j] = solve(s,t,i+1,j,dp);
    }
}

// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna